File: Codes/1574-shortest-subarray-to-be-removed-to-make-array-sorted.py
'''
Author: LetMeFly
Date: 2023-03-25 09:22:32
LastEditors: LetMeFly
LastEditTime: 2023-03-25 09:52:10
'''
from typing import List

class Solution:
    def findLengthOfShortestSubarray(self, arr: List[int]) -> int:
        right = len(arr) - 1
        while right - 1 >= 0 and arr[right - 1] <= arr[right]:
            right -= 1
        if not right:
            return 0
        ans = right
        for left in range(len(arr)):  # left和right肯定不会相遇
            while right < len(arr) and arr[right] < arr[left]:
                right += 1
            ans = min(ans, right - left - 1)
            if left + 1 < len(arr) and arr[left + 1] < arr[left]:
                break
        return ans

File: Codes/1574-shortest-subarray-to-be-removed-to-make-array-sorted.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-03-25 09:22:25
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-03-25 09:44:45
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int findLengthOfShortestSubarray(vector<int>& arr) {
        int right = arr.size() - 1;
        while (right - 1 >= 0 && arr[right - 1] <= arr[right]) {
            right--;
        }
        if (!right) {  // 移动到首个元素了
            return 0;
        }
        int ans = right;
        for (int left = 0; left < arr.size(); left++) {  // left和right肯定不会相遇
            while (right < arr.size() && arr[right] < arr[left]) {
                right++;
            }
            ans = min(ans, right - left - 1);
            // printf("left = %d, right = %d, ans = %d\n", left, right, ans);  //*************
            if (left + 1 < arr.size() && arr[left + 1] < arr[left]) {
                break;
            }
        }
        return ans;
    }
};

#ifdef _WIN32
/*
[10,13,17,21,15,15,9,17,22,22,13]

*/
int main() {
    string s;
    while (cin >> s) {
        vector<int> v = stringToVector(s);
        Solution sol;
        cout << sol.findLengthOfShortestSubarray(v) << endl;
    }
    return 0;
}
#endif